### PR TITLE
Caikun/generator

### PR DESCRIPTION
**Pull request by caikun-pjlab**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation and Context

Functions such as diopiUniform support the generator parameter

## Checklist
**Before PR**:

- [x] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: proto/include/diopi/functions.h [hunk 1]
```diff
@@ -2985,40 +2985,46 @@ DIOPI_API diopiError_t diopiIndexPut(diopiContextHandle_t ctx, diopiTensorHandle
  * @param[in] inout the input and output tensor, type = [float32, float64, float16, int64, int32, int16, int8]
  * @param[in] from the lower bound of the random function. type = [int64].
  * @param[in] to a pointer, the upper bound of the random function, it can be none.
+ * @param[in] generator a pseudorandom number generator for sampling
  */
-DIOPI_API diopiError_t diopiRandomInp(diopiContextHandle_t ctx, diopiTensorHandle_t inout, int64_t from, const int64_t* to);
+DIOPI_API diopiError_t diopiRandomInp(diopiContextHandle_t ctx, diopiTensorHandle_t inout, int64_t from, const int64_t* to,
+                                      diopiConstGeneratorHandle_t generator);
```
Comment (yangbofun) [inline proto/include/diopi/functions.h:2991]: 这个是不是应该是去掉const，直接是diopiGeneratorHandle_t, 下同

File: impl/camb/functions/normal.cpp [hunk 2]
```diff
@@ -29,23 +21,22 @@ diopiError_t diopiNormal(diopiContextHandle_t ctx, diopiTensorHandle_t out, doub
...
-    DiopiTensor state = requiresBuffer(ctx, stateSize);
-
-    uint32_t seed = getSeed();
-    DIOPI_CALLCNNL(cnnlRandMakeMTGP32KernelState(handle, state.data(), nullptr, nullptr, seed));
-    DIOPI_CALLCNNL(cnnlRandGenerateNormal(handle, generator, dtype, state.data(), tensor.numel(), mean, std, tensor.data()));
-    DIOPI_CALLCNNL(cnnlRandDestroyGenerator(generator));
+    DIOPI_CALLCNNL(cnnlRandCreateGenerator(&cnnlGenerator, CNNL_RAND_RNG_MTGP32));
+
+    diopiTensorHandle_t stateHandle = nullptr;
+    DIOPI_CALL(diopiGeneratorGetState(ctx, generator, &stateHandle));
```
Comment (yewentao256) [inline impl/camb/functions/normal.cpp:29]: 这里写的有问题，cnnlRandCreateGenerator会修改cnnl generator，但这里并没有用上

File: proto/include/diopi/functions.h (hunk 1 above)
Comment (caikun-pjlab) [reply proto/include/diopi/functions.h]: 已修改成 diopiGeneratorHandle_t

File: impl/camb/functions/normal.cpp (hunk 2 above)
Comment (caikun-pjlab) [reply impl/camb/functions/normal.cpp:29]: 在 cnnlRandGenerateNormal 里面用到了

File: diopi_test/include/litert.hpp [hunk 3]
```diff
@@ -136,6 +137,20 @@ struct diopiTensor {
     diopiContextHandle_t getCtx() const { return context_; }
 };
 
+struct diopiGenerator {
+private:
+    mutable diopiTensor state_;
```
[the remark this answers is not in the document]
Comment (caikun-pjlab) [reply diopi_test/include/litert.hpp]: 已修改，现在不需要设置成 mutable 了

File: proto/include/diopi/diopirt.h [hunk 4]
```diff
@@ -130,6 +134,9 @@ extern DIOPI_RT_API diopiError_t diopiRequireTensor(diopiContextHandle_t ctx, di
 extern DIOPI_RT_API DIOPI_ATTR_WEEK diopiError_t diopiRequireBuffer(diopiContextHandle_t ctx, diopiTensorHandle_t* tensor, int64_t num_bytes,
                                                                     diopiDevice_t device);
 
+extern DIOPI_RT_API diopiError_t diopiGeneratorGetState(diopiContextHandle_t ctx, diopiConstGeneratorHandle_t th, diopiTensorHandle_t* data);
```
[the remark this answers is not in the document]
Comment (caikun-pjlab) [reply proto/include/diopi/diopirt.h]: 已修改

**Review by LeungChiNan — APPROVED** [review]

File: impl/camb/functions/uniform.cpp [hunk 5]
```diff
@@ -29,18 +22,17 @@ diopiError_t diopiUniformInp(diopiContextHandle_t ctx, diopiTensorHandle_t inout
...
-
-    size_t stateSize = 0;
-    DIOPI_CALLCNNL(cnnlRandGetMTGP32StateSize(generator, &stateSize));
-    DiopiTensor state = requiresBuffer(ctx, stateSize);
-
-    uint32_t seed = getSeed();
-    DIOPI_CALLCNNL(cnnlRandMakeMTGP32KernelState(handle, state.data(), nullptr, nullptr, seed));
-    DIOPI_CALLCNNL(cnnlRandGenerateUniform(handle, generator, dtype, state.data(), tensor.numel(), from, to, tensor.data()));
-    DIOPI_CALLCNNL(cnnlRandDestroyGenerator(generator));
+    DIOPI_CALLCNNL(cnnlRandCreateGenerator(&cnnlGenerator, CNNL_RAND_RNG_MTGP32));
```
Comment (fandaoyi) [inline impl/camb/functions/uniform.cpp:27]: dropout算子也用了随机数。

File: impl/camb/functions/normal.cpp (hunk 2 above)
Comment (fandaoyi) [reply impl/camb/functions/normal.cpp:29]: 类似的问题， 我们的 Generator 没有持有 cnnlRandGenerator_t。  能确定cnnlRandGenerateNormal() 会修改   statePtr 的值吗？ 
 只有这样，diopiGeneratorSetState 才可以保留geneartor 的state吧？ 因为随机数应该是有 seed +state 两部分决定初始状态的。

Comment (fandaoyi) [reply impl/camb/functions/normal.cpp:29]: 好的， 我前面写错了， seed + offset。 看起来 state 可以包含这两部分

File: impl/camb/functions/uniform.cpp (hunk 5 above)
Comment (caikun-pjlab) [reply impl/camb/functions/uniform.cpp:27]: 已适配dropout算子

**Pull request merged**